### PR TITLE
Update user registration flow

### PR DESCRIPTION
**Pull request by mumarkhan999**
This PR is related to [https://edlyio.atlassian.net/browse/EDE-370](https://edlyio.atlassian.net/browse/EDE-370)

**PR Description**
In this PR the regular flow of edX registration is updated. The admin user will upload an admission file, having credentials of students who will be eligible to create an account on the platform,from Django admin. Using the data of the admission file, prospective users will be created. Before this each row of admission file (CSV format) will be validated and if any error has occurred it will be written on the admission file.

The emails which we were sending before upon the registration of new users have been removed. As for now, we are just creating prospective users. After that, the student will register by itself.

The sample output file is given below
image

### CODE REVIEW

File: openedx/features/qverse_features/registration/signals.py [hunk 1]
```diff
@@ -192,9 +134,47 @@ def _validate_single_csv_row(row):
...
+        errors.append('Level ID must be greater than 0 and smaller than {}.'.format(MAX_LEVEL_CHOICES+1))
+
+    if int(row['programmeid']) < 1 or int(row['programmeid']) > MAX_PROGRAMME_CHOICES:
+        errors.append('Programme ID must be greater than 0 and smaller than {}.'.format(MAX_PROGRAMME_CHOICES))
+
+    if errors:
+        # to write multiple lines
+        # in a single cell of csv
+        errors = '\n'.join(errors)
+        errors = '"{}"'.format(errors)
```
Comment (HamzaIbnFarooq) [inline openedx/features/qverse_features/registration/signals.py]:
```suggestion
        errors = '"{}"'.format('\n'.join(errors))
```

File: openedx/core/djangoapps/user_api/api.py [hunk 2]
```diff
@@ -213,8 +213,9 @@ def _apply_third_party_auth_overrides(request, form_desc):
 
 class RegistrationFormFactory(object):
     """HTTP end-points for creating a new user. """
-
-    DEFAULT_FIELDS = ["email", "name", "username", "password"]
+    # To remove full name field from the registration form shown to the user
+    # the "name" field has been removed from DEFAULT_FIELDS
```
Comment (HamzaIbnFarooq) [inline openedx/core/djangoapps/user_api/api.py:217]: Can't we do this through REGISTERATION_FIELDS present in env.json?

**Review by HamzaIbnFarooq — APPROVED** [review]
LGTM

File: openedx/core/djangoapps/user_api/api.py (hunk 2 above)
Comment (mumarkhan999) [reply openedx/core/djangoapps/user_api/api.py:217]: Actually, the **name** field is not included in the fields that can be controlled through  **env.json** 
```
"REGISTRATION_EXTRA_FIELDS": {
        "city": "hidden",
        "confirm_email": "hidden",
        "country": "optional",
        "gender": "optional",
        "goals": "optional",
        "honor_code": "optional",
        "level_of_education": "hidden",
        "mailing_address": "hidden",
        "terms_of_service": "hidden",
        "year_of_birth": "optional"
    },
```

File: openedx/features/qverse_features/registration/signals.py (hunk 1 above)
Comment (mumarkhan999) [reply openedx/features/qverse_features/registration/signals.py]: Done.

File: openedx/features/qverse_features/registration/forms.py [hunk 3]
```diff
@@ -3,16 +3,20 @@
...
-from openedx.features.qverse_features.registration.models import QVerseUserProfile
+from openedx.features.qverse_features.registration.models import QVerseUserProfile, ProspectiveUser
+from student.helpers import AccountValidationError
 
 
 class QVerseUserProfileForm(ModelForm):
     """
     Form to add custom registration fields.
     """
+    def __init__(self, *args, **kwargs):
```
Comment (danialmalik) [inline openedx/features/qverse_features/registration/forms.py]: If we are not doing anything here, can we omit this ?

File: openedx/features/qverse_features/registration/forms.py [hunk 4]
```diff
@@ -22,3 +26,57 @@ class Meta(object):
...
+            },
+            'department': {
+                'required': 'Please select a valid Department'
+            },
+            'programme': {
+                'required': 'Please select the valid Programme that you have been enrolled in.'
+            },
+        }
+        help_texts = {
+            'registration_number': 'Enter your registration number provided from your school.',
```
Comment (danialmalik) [inline openedx/features/qverse_features/registration/forms.py:47]:
```suggestion
            'registration_number': 'Enter your registration number provided by your school.',
```
Just a suggestion. Please ignore if doesn't seem valid.

File: openedx/features/qverse_features/registration/forms.py [hunk 5]
```diff
@@ -22,3 +26,57 @@ class Meta(object):
...
+        """
+        cleaned_data = super(QVerseUserProfileForm, self).clean()
+        try:
+            OPTIONAL_FIELDS = ['mobile_number', 'other_name']
+            prospective_user = ProspectiveUser.objects.get(
+                    registration_number__iexact=cleaned_data['registration_number'])
+            for field_name in cleaned_data.keys():
+                if field_name in OPTIONAL_FIELDS:
+                    continue
+                if str(getattr(prospective_user, field_name)).lower() != str(cleaned_data[field_name]).lower():
```
Comment (danialmalik) [inline openedx/features/qverse_features/registration/forms.py:73]: Do we actually need to use `str` here ? wouldn't the data already be a string?

File: openedx/features/qverse_features/registration/models.py [hunk 6]
```diff
@@ -101,8 +108,43 @@ def save(self, *args, **kwargs):
...
+
+    def __str__(self):
+        return '({}) --- {} {}'.format(self.registration_number, self.first_name, self.surname)
+
+    def save(self, *args, **kwargs):
+        """
+        Make sure that no invalid email is saved in model.
+        """
+        validate_email = EmailValidator()
+        validate_email(self.email)
```
Comment (danialmalik) [inline openedx/features/qverse_features/registration/models.py:149]: Can we use a `clean_email` or `clean` method for the purpose ?

**Review by danialmalik — CHANGES_REQUESTED** [review]

File: openedx/features/qverse_features/registration/forms.py (hunk 3 above)
Comment (mumarkhan999) [reply openedx/features/qverse_features/registration/forms.py]: done.

File: openedx/features/qverse_features/registration/forms.py (hunk 5 above)
Comment (mumarkhan999) [reply openedx/features/qverse_features/registration/forms.py:73]: Nope.
image

File: openedx/features/qverse_features/registration/forms.py (hunk 4 above)
Comment (danialmalik) [reply openedx/features/qverse_features/registration/forms.py:47]: image
This is the actual suggestion that I gave. If github isn't rendering it correctly.

File: openedx/features/qverse_features/registration/models.py (hunk 6 above)
Comment (mumarkhan999) [reply openedx/features/qverse_features/registration/models.py:149]: It didn't work the way I needed it.

**Review by danialmalik — APPROVED** [review]

**Pull request merged**